src/main/scala/io/iohk/ethereum/merklepatriciatrie/Node.scala

```diff
@@ -0,0 +1,102 @@
+package io.iohk.ethereum.merklepatriciatrie
+
+import io.iohk.ethereum.rlp.{decode => decodeRLP, encode => encodeRLP, _}
+import io.iohk.ethereum.merklepatriciatrie.MerklePatriciaTrie.HashFn
+
+/**
+  * Trie elements
+  */
+private[merklepatriciatrie] sealed trait Node {
+
+  import MerklePatriciaTrie._
+
+  private var hashedCache: Option[Array[Byte]] = None
```

Comment:
maybe solution that I proposed will not work her, my main concern here is that we have method and var for something that may be replaced by `lazy val`

about case of decoding, maybe we can have 
```
trait T {val a: Option[hashFn[}
case class Foo(a: Option[hashFn]) extends T
```
? (Just another idea that may work)